### DISCUSSION TITLE
Internal storage for MPIPoolExecutor

### DISCUSSION THREAD

**Discussion by jan-janssen** [mpi4py/mpi4py#370] [Ideas]
I would like to store data or even a python object inside the `MPIPoolExecutor` and then submit functions to interact with this object. For example:
```
from pympipool import Executor

def calc(i, j, k):
    return i + j + k

def init_function():
    return {"j": 4, "k": 3, "l": 2}

with Executor(cores=1, init_function=init_function) as p:
    fs = p.submit(calc, 2, j=5)
    print(fs.result())
```
Here `k` is used from the internal storage so the output is `10`. This code example is based on: 
https://pympipool.readthedocs.io/en/latest/interfaces.html#executor
Which I currently use to identify features I would like to have available in the general `MPIPoolExecutor`. While in this example the overhead of communicating `k` multiple times is minimal, `k` could also reference a several GB large array I would like to analyze. So keeping it in memory can be highly beneficial.

**Comment by dalcinl** [comment]
In `mpi4py.futures`, that's what the `initializer` and `initargs` to `MPIPoolExecutor` are  for. 

Within the initializer function, you can store your data in a global variable using the `global` statement, or directly accessing the `__worker__` namespace to store the global variable.

```python
from mpi4py.futures import MPIPoolExecutor

data = None

def init(arg):
    import __worker__
    __worker__.data = arg

def task(arg):
    return data + arg

if __name__ == '__main__':
    with MPIPoolExecutor(initializer=init, initargs=(40,)) as e:
        print(e.submit(task, 2).result())
```

```shell
$ mpiexec -n 1 python script.py 
42
```
